src/librustc_metadata/encoder.rs

```diff
@@ -194,6 +194,13 @@ impl<'tcx> SpecializedEncoder<Span> for EncodeContext<'tcx> {
             return TAG_INVALID_SPAN.encode(self)
         }
 
+        // HACK(eddyb) there's no way to indicate which crate a Span is coming
+        // from right now, so decoding would fail to find the SourceFile if
+        // it's not local to the crate the Span is found in.
+        if self.source_file_cache.is_imported() {
+            return TAG_INVALID_SPAN.encode(self)
+        }
```

Comment:
cc @petrochenkov I had to add this because decoding assumes `Span`s are always from each crate's own files, not some transitive dependency.

I wonder if this should be an assert instead - I could enforce that the `Span`s computed for `inferred_outlives_of` are always from the definition of the type, not some other type (which is how the last commit in this PR triggered the problem).